src/common/bpf.rs

```diff
@@ -10,6 +10,71 @@ pub struct BPF {
 #[cfg(not(feature = "bpf"))]
 pub struct BPF {}
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+#[cfg(feature = "bpf")]
+pub enum ProbeType {
+    Kernel,
+    User,
+    Tracepoint,
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+#[cfg(feature = "bpf")]
+pub enum ProbeLocation {
+    Entry,
+    Return,
+}
+
+// Define a probe.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+#[cfg(feature = "bpf")]
+pub struct FunctionProbe {
+    pub name: String,                  // name of the function to probe
+    pub handler: String,               // name of the probe handler
+    pub probe_type: ProbeType,         // function type, kernel, user or tracepoint.
+    pub probe_location: ProbeLocation, // probe location, at entry or at return.
+    pub binary_path: Option<String>,   // required for user probe only.
+    pub sub_system: Option<String>,    // required for tracepoint only.
+}
+
+#[cfg(feature = "bpf")]
+impl FunctionProbe {
+    // try to attach to a bpf instance.
+    #[allow(unused_assignments)]
+    pub fn try_attach_to_bpf(&self, bpf: &mut bcc::BPF) -> Result<(), anyhow::Error> {
+        match self.probe_type {
+            ProbeType::Kernel => match self.probe_location {
+                ProbeLocation::Entry => bcc::Kprobe::new()
+                    .handler(self.handler.as_str())
```

Comment:
`.handler(&self.handler)` is more idiomatic

Reply:
thanks for the comment, TIL.